### PR TITLE
refactor: Clean up world saving/loading codepath (Phase 1 of v2 save format)

### PR DESCRIPTION
**Pull request by randombk**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style). => Awaiting CI bot
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

## Purpose of change

This PR is Phase 1 of #5771.

It refactors much of the world and player save read/write logic, with a particular focus on consolidating the business logic of what gets saved where. The purpose is to create a clean environment where we can add on support for different save formats in a subsequent PR.

**This PR should not change the on-disk format. Everything should continue to work normally.** The actual SQLite+compression implementation will happen on top of this refactored backend in a later PR.

## Describe the solution

 * Clean up and refactor `worldfactory` and `world`.
   * The old `WORLD` struct is now renamed `WORLDINFO` to better reflect its actual purpose.
   * Get rid of the `WORLDPTR = WORLD*` convention. This isn't the Windows API, and nothing else in the codebase does this.
 * Move file saving logic into a new class called `world`, meant to encapsulate the idea of a loaded world. All reads and writes to world and player files now go through this class, which forms the foundation for future changes to the save format. 
 * Simplified the I/O utility methods `read_from_file*` and `write_to_file*` found in `cata_utility.cpp`. This replaces the 2 write and 6 read methods with 1 write and 2 read. This simplification is needed because future save loaders will need to provide their own implementations of these methods.
      * The `_optional` variants can be replaced with default arguments.
      * The `JsonDeserializer` variants are never used.
      * The `_json` variants do save some boilerplate (`path`) so I'll keep that in for now.

## Testing

Awaiting CI signals.
I tested by loading & saving a couple of existing worlds, without obvious issues thus far.

### CODE REVIEW

**Comment by scarf005** [comment]
gah astyle is going haywire

**Comment by randombk** [comment]
No, `astyle` is largely correct here. I'll pull and squash. I don't have a proper astyle setup locally so I'm just depending on the bot for now.

**Comment by randombk** [comment]
Actually, if there's no objections I'll  the asyle commit without squashing - it leaves the individual diffs nice and self-consistent.

**Review by scarf005 — DISMISSED** [review]
the changes looks good to me, there's some nitpicks here and there but not critical.

@chaosvolt @RoyalFox2140 could you playtest them before merging to be extra sure?

**Review by joveeater — APPROVED** [review]
Looks good to me. Great stuff.

File: src/game.cpp [hunk 1]
```diff
@@ -2732,16 +2727,23 @@ spell_events &game::spell_events_subscriber()
     return *spell_events_ptr;
 }
 
-static bool save_uistate_data( const game &g )
+bool game::save_uistate_data()
```
Comment (Coolthulhu) [inline src/game.cpp:2730]: Does it affect the state? Should be const if it doesn't

File: src/game.cpp [hunk 2]
```diff
@@ -11413,12 +11418,12 @@ void game::quicksave()
 
 void game::quickload()
 {
-    const WORLDPTR active_world = world_generator->active_world;
+    world *active_world = get_active_world();
```
Comment (Coolthulhu) [inline src/game.cpp:11421]: Does it have to be non-const?

File: src/world.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    //            "compression    TEXT DEFAULT NULL," \
+    //            "data           BLOB NOT NULL" \
+    //            ");";
+
+
+    // char *sqlErrMsg = 0;
+    // if( SQLITE_OK != ( ret = sqlite3_exec( db, sql, NULL, NULL, &sqlErrMsg ) ) ) {
+    //     dbg( DL::Error ) << "Failed to init db" << db_path << " (" << sqlErrMsg << ")";
+    //     throw std::runtime_error( "Failed to open db" );
+    // }
```
Comment (Coolthulhu) [inline src/world.cpp:256]: Commented out code - should be removed from the PR. It's easy to restore it from git history later.

File: src/cata_utility.cpp [hunk 4]
```diff
@@ -523,8 +530,12 @@ std::istream &safe_getline( std::istream &ins, std::string &str )
     }
 }
 
-bool read_from_file( const std::string &path, const std::function<void( std::istream & )> &reader )
+bool read_from_file( const std::string &path, file_read_cb reader, bool optional )
```
Comment (Coolthulhu) [inline src/cata_utility.cpp:533]: Yeah, that `bool` is hard to understand there. Especially after function definition.
If not an enum flag, the following options would also work:
* Split into 2 functions: mandatory and optional variant
* Empty class used just as argument here, for overload

**Comment by scarf005** [comment]
> I'd like to see if the various code nitpicks can be addressed before I test it. Are any of them deal-breakers or are they all just minor things that can be ignored if nessecary?

they are pure C++ nitpicks and won't affect gameplay.

**Comment by chaosvolt** [comment]
In that case will see about testing it in a bit, later today if getting ready to head out and deal with errands doesn't leave be with enough time.

**Review by chaosvolt — APPROVED** [review]
1. Compile-tested.
2. Loaded a save from a previous compile-test, no errors seen.
3. Started a new world, saved and loaded, also worked fine.
4. Started a new character in this new world with one already in it, saving and loading the second character also works.

**Comment by randombk** [comment]
Addressed comments. The consts were a leftover from a previous iteration - I added them back in where appropriate.

The player data write methods cannot be const to make way for the Phase 2 patch. Because the saveid is not known when the `world` object is created, we must load the player sqlite file on-demand thus requiring some internal mutable state during player writes.

**Comment by scarf005** [comment]
strange, MSVC is chocking on here
```
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2510,63): error C2589: ')': illegal token on right side of '::' [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2510): error C2062: type 'unknown-type' unexpected [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2564,117): error C2589: ')': illegal token on right side of '::' [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2564): error C2062: type 'unknown-type' unexpected [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2590,30): error C2589: ')': illegal token on right side of '::' [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
D:\a\Cataclysm-BN\Cataclysm-BN\src\game.cpp(2590): error C2062: type 'unknown-type' unexpected [D:\a\Cataclysm-BN\Cataclysm-BN\out\build\windows-tiles-sounds-x64-msvc\src\cataclysm-bn-tiles-common.vcxproj]
```

**Comment by randombk** [comment]
Yea, there's something about enum syntax that MSVC really doesn't like.

**Comment by scarf005** [comment]
weird, other enum classes compile fine.

**Comment by RoyalFox2140** [comment]
> the changes looks good to me, there's some nitpicks here and there but not critical.
> 
> @chaosvolt @RoyalFox2140 could you playtest them before merging to be extra sure?

If the saves are going to break they're going to break after days of playing the same save file for hours a day, And I can't play self-compile versions. No matter what settings are used my game lags heavily and we've gone over using Release without Debug symbols doing nothing to fix it, and proper releases not lagging. There's nothing for me to do here. If it's going to break everything it's going to do so in a way that won't get caught from standard testing.

MSVC and Ubuntu tests failed, so I'll at least re-run those.

**Comment by randombk** [comment]
I changed the enum back to a bool. There's something weird going on with MSVC and I don't have a local environment to test/debug that right now.

**Review by chaosvolt — APPROVED** [review]
1. Compiled and load-tested.
2. Loaded a save from previous compiling, loads as expected. Loading the two different characters in the save also work right.
3. Tested save/load, no oddities found.
4. Also quicksaved, debug quit to menu, then loaded with no issues.
5. Spawned in an APC and used autodrive to move a long distance before quicksaving, also worked.

Quicksave just standing in starting shelter:
image

Quicksave after a nice long drive (about 75 thousand blocks of things to save, whatever the number during save progress is counting):
image

**Pull request merged**